std-bits/table/src/main/java/org/enso/table/read/DelimitedReader.java

```diff
@@ -203,85 +198,140 @@ private void reportProblem(ParsingProblem problem) {
+    if (cell == null) return false;
+    Object parsed = cellTypeGuesser.parseSingleValue(cell, noOpProblemAggregator);
+    return parsed instanceof String;
+  }
+
   /** Reads the input stream and returns a Table. */
   public Table read() {
     List<String> headerNames;
-    String[] currentRow = nextRow();
+    String[] currentRow = readNextRow();
 
     // Skip the first N rows.
     for (long i = 0; currentRow != null && i < skipRows; ++i) {
-      currentRow = nextRow();
+      currentRow = readNextRow();
     }
 
     // If there are no rows to even infer the headers, we return an empty table.
     if (currentRow == null) {
       return new Table(new Column[0]);
     }
 
-    switch (headerBehavior) {
-      case INFER:
-        throw new IllegalStateException("Inferring headers is not yet implemented");
-      case USE_FIRST_ROW_AS_HEADERS:
-        List<String> preprocessedHeaders =
-            Arrays.stream(currentRow).map(this::parseHeader).collect(Collectors.toList());
-        headerNames = NameDeduplicator.deduplicate(preprocessedHeaders, "_");
-        // We have 'used up' the first row, so we load a next one.
-        currentRow = nextRow();
-        break;
-      case GENERATE_HEADERS:
-        headerNames = new ArrayList<>(currentRow.length);
-        for (int i = 0; i < currentRow.length; ++i) {
-          headerNames.add(COLUMN_NAME + "_" + (i + 1));
-        }
-        break;
-      default:
-        throw new IllegalStateException("Impossible branch.");
-    }
-
-    StringStorageBuilder[] builders = initBuilders(headerNames.size());
-
-    while (currentRow != null && (rowLimit < 0 || target_table_index < rowLimit)) {
-      if (currentRow.length != builders.length) {
-        reportInvalidRow(current_line, keepInvalidRows ? target_table_index : null, currentRow);
+    int expectedColumnCount = currentRow.length;
+    initBuilders(expectedColumnCount);
 
-        if (keepInvalidRows) {
-          for (int i = 0; i < builders.length && i < currentRow.length; i++) {
-            builders[i] = builders[i].parseAndAppend(currentRow[i]);
-          }
-
-          // If the current row had less columns than expected, nulls are inserted for the missing
-          // values.
-          // If it had more columns, the excess columns are discarded.
-          for (int i = currentRow.length; i < builders.length; i++) {
-            builders[i] = builders[i].parseAndAppend(null);
+    assert currentRow != null;
+    switch (headerBehavior) {
+      case INFER -> {
+        String[] firstRow = currentRow;
+        String[] secondRow = readNextRow();
+        if (secondRow == null) {
+          // If there is only one row in the file, we generate the headers and stop further processing (as nothing more to process).
+          headerNames = generateDefaultHeaders(expectedColumnCount);
+          appendRowIfLimitPermits(firstRow);
+          currentRow = null;
+        } else {
+          assert cellTypeGuesser != null;
```

Comment:
is this assert necessary? I'm generally not a fan of adding so many of these as they obscure the logic. I would instead rely on tests

Reply:
What do you mean by "necessary"? I mean, I do not need to add any assertions at all. I just wanted to have it here to express that we are sure that `cellTypeGuesser` is not `null` here. It is quite obvious because we have the `if` check above, so even the IDE says that the assert can be simplified to just `assert true`. But I like to keep it, in case some logic above changes, this would allow us to detect any issues. What do you think?

Reply:
Sure, it can stay. My preference is to bubble runtime exceptions if some logic is seriously experimental instead of adding assertions in prod code. Better optics for users IMHO as well. But it is a personal preference so not gonna insist on that.